compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/TypeChecker.java

```diff
@@ -9918,6 +9914,24 @@ String recordsToString(Set<BRecordType> recordTypeSet) {
         }
     }
 
+    public GlobalStateData getGlobalStateSnapshotAndResetGlobalState() {
```

Comment:
Can't we do this via analyzer data instead of introducing a new data holder?

Reply:
`AnalyzerData` act like set of global fields where we pass through the functions. `GlobalStateData` is used as a snapshot of several fields where we will be creating multiple of objects of it. However, we can still use the `AnalyzerData` for the same purpose. The issue is it will introduce unwanted object creations of the `AnalyzerData` which is not clean

Reply:
The addition is basically resetting unknown type refs, right? Can we just do it inline instead of introducing `GlobalStateData`? The name is ambiguous and can be confusing with analyzer data IMO.

Reply:
Writing this inline is not a good idea in terms of scalability. We already have two fields that needs to be reset. There is a high chance of more fields require this in future. I have renamed the class name to `GlobalStateSnapshot`

Reply:
The only additional thing we've done is changing unknown type refs, right? Error count logic was already there? It makes sense to extract out repeated logic to a function, but IMO, here we are just doing it for part of it. I would rather refactor the whole thing properly or just reset type refs in line (similar to the others). With just what we have now, I also don't think GlobalStateSnapshot is an ideal name here, since it doesn't include all "global state" and can be confusing with analyzer data.

I would introduce the change in a way that makes it readable and clear at the moment, rather than anticipating future changes (when we can refactor everything properly).

Reply:
I don't think this approach is less readable than the inline one. If we write this inline the code becomes messy, not scalable, and the same code will duplicate at every place where we do silent type checking. 

> It makes sense to extract out repeated logic to a function

Yes, to extract out repeated logic we need some kind of a object or  a record to get the previous data available at the caller function